packages/server-wallet/src/wallet/index.ts

```diff
@@ -0,0 +1,238 @@
+            const addresses = ss.participants.map(p => p.signingAddress);
+            const signingWallet = await SigningWallet.query(tx)
+              .whereIn('address', addresses)
+              .first();
+
+            if (!signingWallet) {
+              logger.error(
+                {
+                  knownWallets: await SigningWallet.query(tx).select(),
+                  addresses,
+                },
+                'Not in channel'
+              );
+              throw Error('Not in channel');
+            }
+
+            const {address: signingAddress} = signingWallet;
+            const cols: RequiredColumns = {
+              ...ss,
+              vars: [addHash(ss)],
+              signingAddress,
+            };
+
+            channel = Channel.fromJson(cols);
+            const {channelId} = await Channel.query(tx).insert(channel);
+            channelIds.push(channelId);
+          } else {
+            ss.signatures?.map(sig => channel.addState(ss, sig));
+            await Channel.query(tx).update(channel);
+            channelIds.push(channel.channelId);
+          }
+        }
+      });
+    } catch (err) {
+      logger.error({err}, 'Could not push message');
+      throw err;
+    }
+
+    const {outbox} = await protocolEngine(channelIds);
+
+    return {
+      channelResults: [
+        {
+          outbox,
+          appData: '',
+          appDefinition: '',
+          channelId: '',
+          challengeExpirationTime: 0,
+          status: 'funding',
+          participants: [],
+          allocations: [],
+          turnNum: 0,
+        },
+      ],
+    };
+  }
+  onNotification(_cb: (notice: Notification) => void): {unsubscribe: () => void} {
+    throw 'Unimplemented';
+  }
+}
+
+type ExecutionResult = {ids: Bytes32[]; outbox: Outgoing[]};
+const protocolEngine = async (
+  ids: Bytes32[],
+  outbox: Outgoing[] = []
+): Promise<ExecutionResult> => {
+  if (ids.length === 0) return {ids, outbox};
+
+  const [channelId, ...nextIds] = ids;
+
+  const channel = await Channel.forId(channelId, undefined);
+  const actions = await executionLoop(channel);
+  const todos = actions.filter(isInternal);
+
+  for (const todo of todos) {
+    let state: SignedState;
+    switch (todo.type) {
+      case 'SignState':
+        state = await channel.signState(todo.hash);
+        break;
+      case 'UpdateChannel':
+        // channel.update(todo);
+        break;
+    }
+
+    await Channel.query().update(channel);
+    outbox.push({
+      type: 'NotifyApp',
+      notice: {
+        method: 'MessageQueued',
+        params: {
+          recipient: 'bob',
+          sender: 'alice',
+          data: {signedStates: [state]},
+        },
+      },
+    });
+
+    // TODO: Push the channelId back, once we actually trigger these actions
+    // nextIds.push(channelId);
+  }
+
+  outbox = outbox.concat(actions.filter(isOutgoing));
+
+  if (nextIds.length) {
+    return await protocolEngine(nextIds, outbox);
```

Comment:
I think it might be marginally more descriptive to write this in a loop format instead of using recursion. Something like the following seems to track the algorithm we're following pretty closely:
```ts
let outbox = [];
let channelIdQueue = [initialChannelId];
while (channelIdQueue.length > 0) {
    const [channelId, ...restOfQueue] = channelIdQueue;
    const [messages, updatedChannelIds] = executeLoop(channelId);
    outbox = [...outbox, ...messages];
    channelIdQueue = [...restOfQueue, ...updatedChannelIds];    //TODO: remove duplicates
}
```

Reply:
Also wondering if `executeLoop` should be `executeLoopStep` or `attemptToAdvanceState` or `takeNextAction` or something.